link-grammar/disjunct-utils.c

```diff
@@ -101,13 +101,17 @@ static inline unsigned int old_hash_disjunct(disjunct_dup_table *dt, Disjunct *
 	unsigned int i;
 	i = 0;
 	for (e = d->left ; e != NULL; e = e->next) {
-		i += ((uintptr_t)e->desc) + e->desc->uc_num;
+		i = (5 * (i + e->desc->uc_num)) + (unsigned int)e->desc->lc_letters + 7;
 	}
 	for (e = d->right ; e != NULL; e = e->next) {
-		i += ((uintptr_t)e->desc) + e->desc->uc_num;
+		i = (5 * (i + e->desc->uc_num)) + (unsigned int)e->desc->lc_letters + 7;
```

Comment:
Isn't it an FNV-hash-like?

``` text
   hash = FNV_offset_basis
   for each byte_of_data to be hashed
        hash = hash × FNV_prime
        hash = hash XOR byte_of_data
   return hash
```
The prime I used is 5, and I used addition (the same overhead of XOR).
BTW multiplying by 5 is very fast (one load instruction on x86_64).
I tried higher prime values but didn't get a collision improvement (I didn't try XOR, but I don't expect a noticeable improvement in this specific case). However, an addition of a  constant gave a noticeable improvement (this is not general, maybe because of very few connectors and with low connector variation), and I tried several values until settled on 7.

Reply:
Yeah. I think the XOR does some scrambling that addition doesn't.  Addition makes things Abelian.  I dunno, maybe this is fine in the current implementation.